connlimit.go

```diff
@@ -182,11 +193,26 @@ func (l *Limiter) SetConfig(c Config) {
 // their own calls if they need to continue limiting the number of concurrent
 // hijacked connections.
 func (l *Limiter) HTTPConnStateFunc() func(net.Conn, http.ConnState) {
+	message := "Your IP is issuing too many concurrent connections, please rate limit your calls\n"
+	tooManyRequestsResponse := []byte(fmt.Sprintf("HTTP/1.1 429 Too Many Requests\r\n"+
+		"Content-Type: text/plain\r\n"+
+		"Content-Length: %d\r\n"+
+		"Connection: close\r\n\r\n%s", len(message), message))
 	return func(conn net.Conn, state http.ConnState) {
 		switch state {
 		case http.StateNew:
 			_, err := l.Accept(conn)
 			if err != nil {
+				if err == ErrPerClientIPLimitReached {
+					go func() {
+						// We don't care about slow players
+						if l.RWDeadlineMaxDelay > 0 {
+							conn.SetWriteDeadline(time.Now().Add(l.RWDeadlineMaxDelay))
+						}
+						conn.Write(tooManyRequestsResponse)
+						conn.Close()
+					}()
+				}
 				conn.Close()
```

Comment:
This close races with the Write in the goroutine, so it is indeterminate whether the response will get written. Moving *this* Close to an `else` means the connection may be handled before the closing goroutine runs.

Reply:
yes, I made my change too fast, there used to be a return (and what I tested in Consul codebase)